Add Raspberry Pi 500's RP2040 board support Fixes #2640

jhordies opened the pull request:
This PR adds board support for the Raspberry Pi 500's embedded RP2040 microcontroller, enabling custom firmware development while maintaining compatibility with the original keyboard functionality.

**Changes:**

Add raspberry_pi_pi500.h board definition
Configure W25X10CL flash support
Document complete GPIO pin mapping for 8×18 keyboard matrix
Define system function pins (power control, LEDs, debug UART)
Enable USB operation independent of main Pi power state

**Hardware Specifications:**

Chip: RP2040 (same as Pico)
Flash: W25Q16JVUXIQ 2MB (DSPI mode)
USB: Connected to Pi 500 USB hub
Matrix: 8 rows × 18 columns keyboard matrix
System GPIO: GP16 (UART),  GP19 (power), GP20 (power key), GP25 (Caps LED)

**Testing:**
✅ Successful firmware compilation with
✅ USB device enumeration and communication
✅ System function compatibility (power, LEDs)
✅ Compatible with original keyboard firmware restoration

**Use Cases:**
This enables Pi 500 users to develop custom USB applications like MIDI controllers, HID devices, and keyboard firmware modifications while preserving hardware compatibility. Note that unlike standard Pico boards, no GPIO pins are available for general use.

**Files Changed:**
src/boards/include/boards/raspberry_pi_pi500_rp2040.h (new)

**Usage:**
-DBOARD=raspberry_pi_pi500_rp2040

**⚠️ CRITICAL WARNING**: Custom firmware must implement power button handling (GP19 control) or the Pi 500 power button will stop working, potentially making the device unbootable. Always include power management code from the original QMK firmware.

jhordiesluminus committed:
Add Raspberry Pi 500 board support

- W25X10CL flash configuration for DSPI mode
- USB standalone operation support
- Complete GPIO pin documentation with keyboard matrix mapping
- Debug UART configuration on GP16
- Hardware-specific boot and power management settings

Tested with TinyUSB MIDI example - successful enumeration and operation.

lurch commented:
As I attempted to hint at in the linked issue, the Raspberry Pi 500 keyboard uses RP2040, not RP2350.

Also, IMHO it would be better if this header file just declared everything explicitly, rather than doing
```c
#include "boards/pico2.h"
```
And the filename should perhaps clarify that this is only the **keyboard** part of the Pi 500, not the entire thing :laughing:

jhordiesluminus committed:
Correct Pi 500 board support: RP2040 chip, proper naming

- Rename to raspberry_pi_pi500_rp2040.h for clarity
- Correct chip type from RP2350 to RP2040
- Remove pico2.h inheritance, use self-contained definitions
- Add comprehensive GPIO pin mapping documentation
- Include critical power management warning
- Configure W25X10CL flash support for RP2040

lurch commented on src/boards/include/boards/raspberry_pi_pi500_rp2040.h:
```diff
@@ -0,0 +1,127 @@
+#ifndef PICO_DEFAULT_UART_TX_PIN
+#define PICO_DEFAULT_UART_TX_PIN 16  // Pi 500 debug UART on GP16
+#endif
+#ifndef PICO_DEFAULT_UART_RX_PIN
+#define PICO_DEFAULT_UART_RX_PIN 1   // Standard RX pin
+#endif
```
According to your comments below, GP1 is used as a matrix row, so you shouldn't define it here.

lurch commented on src/boards/include/boards/raspberry_pi_pi500_rp2040.h:
```diff
@@ -0,0 +1,127 @@
+#define PICO_DEFAULT_LED_PIN 17  // Pi 500 LED on GP17
+#endif
+
+// --- I2C ---
+#ifndef PICO_DEFAULT_I2C
+#define PICO_DEFAULT_I2C 0
+#endif
+#ifndef PICO_DEFAULT_I2C_SDA_PIN
+#define PICO_DEFAULT_I2C_SDA_PIN 4
+#endif
+#ifndef PICO_DEFAULT_I2C_SCL_PIN
+#define PICO_DEFAULT_I2C_SCL_PIN 5
+#endif
+
+// --- SPI ---
+#ifndef PICO_DEFAULT_SPI
+#define PICO_DEFAULT_SPI 0
+#endif
+#ifndef PICO_DEFAULT_SPI_SCK_PIN
+#define PICO_DEFAULT_SPI_SCK_PIN 18
+#endif
+#ifndef PICO_DEFAULT_SPI_TX_PIN
+#define PICO_DEFAULT_SPI_TX_PIN 19
+#endif
+#ifndef PICO_DEFAULT_SPI_RX_PIN
+#define PICO_DEFAULT_SPI_RX_PIN 16
+#endif
+#ifndef PICO_DEFAULT_SPI_CSN_PIN
+#define PICO_DEFAULT_SPI_CSN_PIN 17
+#endif
```
IMHO it makes **no** sense to define any `PICO_DEFAULT_I2C` or `PICO_DEFAULT_SPI` settings here.

lurch commented on src/boards/include/boards/raspberry_pi_pi500_rp2040.h:
```diff
@@ -0,0 +1,127 @@
+#endif
+
+// Drive high to force power supply into PWM mode (lower ripple on 3V3 at light loads)
+#define PICO_SMPS_MODE_PIN 23
```
Your comments say that this is used as a matrix row?

lurch commented on src/boards/include/boards/raspberry_pi_pi500_rp2040.h:
```diff
@@ -0,0 +1,127 @@
+#define PICO_SMPS_MODE_PIN 23
+
+#ifndef PICO_RP2040_B0_SUPPORTED
+#define PICO_RP2040_B0_SUPPORTED 1
```
This can be changed to 0, since all Pi 500s are using RP2040 B2.

lurch commented on src/boards/include/boards/raspberry_pi_pi500_rp2040.h:
```diff
@@ -0,0 +1,127 @@
+#define PICO_RP2040_B0_SUPPORTED 1
+#endif
+
+// The GPIO Pin used to read VBUS to determine if the device is battery powered.
+#ifndef PICO_VBUS_PIN
+#define PICO_VBUS_PIN 24
+#endif
+
+// The GPIO Pin used to monitor VSYS. Typically you would use this with ADC.
+#ifndef PICO_VSYS_PIN
+#define PICO_VSYS_PIN 29
+#endif
```
As before, you shouldn't be defining things if those pins are actually used by the keyboard matrix.

lurch commented on src/boards/include/boards/raspberry_pi_pi500_rp2040.h:
```diff
@@ -0,0 +1,127 @@
+// --- USB CONFIGURATION ---
+// Crucial for Pi 500 keyboard function when Pi is off - ignores USB startup check
+#define PICO_STDIO_USB_ENABLE_RESET_VIA_VENDOR_INTERFACE 0
+#define PICO_STDIO_USB_RESET_BOOTSEL_ACTIVITY_LED 0
```
I believe this doesn't do what you think it does - it's not disabling USB_RESET_BOOTSEL_ACTIVITY_LED, but setting it to GP0.

lurch commented on src/boards/include/boards/raspberry_pi_pi500_rp2040.h:
```diff
@@ -0,0 +1,127 @@
+
+// --- BOOTLOADER CONFIGURATION ---
+// Disable double tap reset timeout for Pi 500
+#define PICO_BOOTSEL_VIA_DOUBLE_RESET_ACTIVITY_LED 0
```
See previous comment :slightly_smiling_face: 
(and look also at Chapter 6 of https://datasheets.raspberrypi.com/pico/raspberry-pi-pico-c-sdk.pdf )

lurch commented on src/boards/include/boards/raspberry_pi_pi500_rp2040.h:
```diff
@@ -0,0 +1,127 @@
+
+// --- PI 500 KEYBOARD MATRIX CONFIGURATION ---
+// Complete GPIO pin mapping for 8×18 keyboard matrix
+// Rows (8): GP0, GP1, GP2, GP3, GP4, GP5, GP6, GP7
+// Cols (18): GP27, GP8, GP9, GP10, GP11, GP12, GP13, GP14, GP23, GP24, GP22, GP20, GP29, GP18, GP15, GP21, GP28, GP26
```
I guess it _might_ be useful to do e.g.
```c
#define PI500_RP2040_MATRIX_ROW_0_PIN 0
#define PI500_RP2040_MATRIX_ROW_1_PIN 1

#define PI500_RP2040_MATRIX_COL_0_PIN 27
#define PI500_RP2040_MATRIX_COL_1_PIN 8
```
etc. ?

jhordies commented:
> As I attempted to hint at in the linked issue, the Raspberry Pi 500 keyboard uses RP2040, not RP2350.
> 
> Also, IMHO it would be better if this header file just declared everything explicitly, rather than doing
> 
> ```c
> #include "boards/pico2.h"
> ```
> 
> And the filename should perhaps clarify that this is only the **keyboard** part of the Pi 500, not the entire thing 😆

Sorry I didn't get the hint. 
You are correct, and updated the code accordingly.

jhordiesluminus committed:
PICO_RP2040_B0_SUPPORTED can be changed to 0, since all Pi 500s are using RP2040 B2.

lurch commented:
@jhordies Do you have any interest in continuing to push this forwards?

jhordiesluminus committed:
Update Raspberry Pi 500 board support for minimal original intent.
- Fix mistake done in pico.h PICO_RP2040_B0_SUPPORTED 0 => 1
- Disable B0 support in pi500_rp2040
- Fix flash size 1M-Bit not 1Mb
- Add Matrix pin definition

lurch commented:
Thanks! I'll take a look at this next week.

jhordies commented:
> Thanks! I'll take a look at this next week.

Thanks, I tried to make a minimal version integrating your comments

lurch commented:
If you wanted to keep "System GPIO: GP16 (UART)" in this header-file (i.e. `PICO_DEFAULT_UART_TX_PIN` and `PICO_DEFAULT_UART`) I wouldn't mind :wink:

egycm3 reviewed (changes requested):
A few comments from me:

- W25X10CL boot stage 2 flash option is correct, but because only DSPI pins are connected to the RP2040 in Pi500 (the flash chip used is the same as Pico 1 - W25Q16JVUXIQ)
- PICO_FLASH_SIZE_BYTES should be (2 * 1024 * 1024)
- GPIO17 is NC on Pi500 - IIRC it requires a hand mod to add the LED to that pin, so I don't think it should count as a 'default led pin'

jhordiesluminus committed:
Fix flash size.
Add warning on GP19
Remove GP17

jhordies commented:
> * W25Q16JVUXIQ

jhordiesluminus committed:
Adding UART TX GP16

jhordies commented:
Sorry for the delay and closing the PR by mistake.
I tried to reflect your comments in the code.

lurch commented on src/boards/include/boards/raspberry_pi_pi500_rp2040.h:
```diff
@@ -0,0 +1,82 @@
+#endif
+
+// Keyboard indicator LEDs
+#define PI500_RP2040_CAPS_LOCK_LED_PIN 25
```
Perhaps it'd be useful to add:
```c
#ifndef PICO_DEFAULT_LED_PIN
#define PICO_DEFAULT_LED_PIN PI500_RP2040_CAPS_LOCK_LED_PIN
#endif
```
?

jhordiesluminus committed:
Add Caps led as default led

lurch commented:
Thanks for making all the requested tweaks @jhordies  :+1: 
I'm now happy that this is technically correct, but it's up to @peterharperuk and @kilograham to decide if we'll actually be merging this.

jhordies commented:
Thanks @lurch

jhordies commented:
Raspberry is launching the 500+
Their press release seem to indicate they anticipate customer to want to use the micro controller for something else than originally intended. 
[https://www.raspberrypi.com/news/the-ultimate-all-in-one-pc-raspberry-pi-500-plus-on-sale-now-at-200/](https://www.raspberrypi.com/news/the-ultimate-all-in-one-pc-raspberry-pi-500-plus-on-sale-now-at-200/)
_The most obvious physical difference between Raspberry Pi 500+ and 500 is the keyboard. Each key rests on a Gateron KS-33 Blue switch with a custom RAL 7001 Silver Grey stem, giving a satisfying sound and feel when pressed. Individually addressable RGB LEDs provide programmable backlighting, and **with an RP2040 running QMK as the controller, a Doom port to the keyboard itself is surely just a matter of time**_.

lurch commented:
> Their press release seem to indicate they anticipate customer to want to use the micro controller for something else than originally intended.

That's joking about displaying Doom on the keyboard's RGB LEDs, just like the [Flappy Bird game](https://www.raspberrypi.com/documentation/computers/keyboard-computers.html#flappy-bird), and a reference to the fact that RP2040 is powerful enough to [play Doom](https://www.raspberrypi.com/news/doom-comes-to-raspberry-pi-pico/) :wink:
